File: Samples/app/src/main/java/samples/linhtruong/com/app/uireactive/component/AppComponent.java
package samples.linhtruong.com.app.uireactive.component;

import android.content.Context;

import javax.inject.Singleton;

import dagger.Component;
import samples.linhtruong.com.app.uireactive.module.AppModule;
import samples.linhtruong.com.app.uireactive.task.TaskManager;
import samples.linhtruong.com.base.BaseComponent;
import samples.linhtruong.com.base.task.TaskResource;

/**
 * Created by Truong on 9/26/16 - 17:28.
 * Description:
 */

@Singleton
@Component (
        modules = {AppModule.class}
)
public interface AppComponent extends BaseComponent {

    final class Initializer {
        public static AppComponent init(AppModule module) {
            return DaggerAppComponent.builder().appModule(module).build();
        }
    }

    void inject(TaskManager taskManager);
    void inject(TaskResource taskResource);

    Context getContext();

    TaskResource getTaskResource();
    TaskManager getTaskManager();
}

File: BaseProject/app/src/main/java/com/linhtruong/sample/core/di/AppComponent.kt
package news.agoda.com.sample.di.module.di

import android.app.Application
import com.google.gson.Gson
import dagger.Component
import com.linhtruong.sample.core.platform.App
import com.linhtruong.sample.core.di.viewmodel.ViewModelModule
import news.agoda.com.sample.di.module.AppModule
import news.agoda.com.sample.di.module.NetworkModule
import news.agoda.com.sample.di.module.RepositoryModule
import com.linhtruong.sample.explore.NewsDetailFragment
import com.linhtruong.sample.explore.NewsListFragment
import javax.inject.Singleton


/**
 * @author linhtruong
 */
@Singleton
@Component(modules = [AppModule::class, NetworkModule::class, RepositoryModule::class, ViewModelModule::class])
interface AppComponent {
    companion object Initializer {
        fun init(app: App) : AppComponent {
            return DaggerAppComponent.builder().appModule(AppModule(app)).build()
        }
    }

    fun application(): Application
    fun gson(): Gson

    fun inject(app: App)
    fun inject(fragment: NewsListFragment)
    fun inject(fragment: NewsDetailFragment)
}